src/ORM/EagerLoadedList.php

```diff
@@ -0,0 +1,1003 @@
+    {
+        // Throw exception for compatability with DataList
+        throw new BadMethodCallException("Can't alter items in an EagerLoadedList using array-access");
+    }
+
+    /**
+     * Unset an item with the key in $key
+     * @throws BadMethodCallException
+     */
+    public function offsetUnset(mixed $key): void
+    {
+        // Throw exception for compatability with DataList
+        throw new BadMethodCallException("Can't alter items in an EagerLoadedList using array-access");
+    }
+
+    public function count(): int
+    {
+        return count($this->getFinalisedRows());
+    }
+
+    /**
+     * Return the maximum value of the given field in this list
+     *
+     * @param string $fieldName
+     */
+    public function max($fieldName): mixed
+    {
+        return max($this->column($fieldName));
+    }
+
+    /**
+     * Return the minimum value of the given field in this list
+     *
+     * @param string $fieldName
+     */
+    public function min($fieldName): mixed
+    {
+        return min($this->column($fieldName));
+    }
+
+    /**
+     * Return the average value of the given field in this list
+     *
+     * @param string $fieldName
+     */
+    public function avg($fieldName): mixed
+    {
+        // We have to rely on the database to either give us the right answer or throw the right exception.
+        // MySQL does wiether things with sum, e.g. an average for the Text field "1_1" is 1 - but
+        // other database implementations could behave differently.
+        $list = $this->foreignID ? $this->dataList->forForeignID($this->foreignID) : $this->dataList;
+        return $list->byIDs($this->column('ID'))->avg($fieldName);
+    }
+
+    /**
+     * Return the sum of the values of the given field in this list
+     *
+     * @param string $fieldName
+     */
+    public function sum($fieldName): int|float
+    {
+        // We have to rely on the database to either give us the right answer or throw the right exception.
+        // MySQL does wiether things with sum, e.g. a sum for the Text field "1_1" is 1 - but
+        // other database implementations could behave differently.
+        $list = $this->foreignID ? $this->dataList->forForeignID($this->foreignID) : $this->dataList;
+        return $list->byIDs($this->column('ID'))->sum($fieldName);
+    }
+
+    /**
+     * Returns true if this list has items
+     */
+    public function exists(): bool
+    {
+        return $this->count() !== 0;
+    }
+
+    public function canFilterBy($fieldName): bool
+    {
+        if (!is_string($fieldName) || empty($this->rows)) {
+            return false;
+        }
+
+        $id = array_key_first($this->rows);
+        return array_key_exists($fieldName, $this->rows[$id]);
+    }
+
+    public function canSortBy($fieldName): bool
+    {
+        return $this->canFilterBy($fieldName);
+    }
+
+    public function find($key, $value): ?DataObject
```

Comment:
```suggestion
    public function find(string $key, string $value): ?DataObject
```
We should strongly type all the things

While I can see why you didn't strongly type lots of args to match DataList, in this instance as well as many others you've added a strong return type which DataList doesn't have.  I think we should just strong type everything here as it's best practice.  Realistically getting the SS_List properly sorted so all the 3x list types match is a CMS 6 thing and that will involve adding strong typing to everything.

Reply:
> We should strongly type all the things

I disagree, and I _intentionally_ didn't add strong typing for parameters that _aren't_ strongly typed in `DataList`. If someone can pass a value into a method on `DataList`, they should be able to pass it in here as well. The specific example you've recommended changing here is a _great_ example - the `$value` might not be a string, it could be null or int or possibly other things.
I don't think we should strongly type these parameters until we're able to commit to strongly typing them in `DataList` as well, so that the types will _always_ be valid or invalid across the board, and give the same errors/exceptions/experience for the developer.

Strong typing for return types just means that the _implementation_ of the method will return the same type, but doesn't affect what people _pass in_ which is the important part.